File: ruby/triangle/triangle.rb
class Triangle
  attr_reader :a, :b, :c

  def initialize(sides)
    @a, @b, @c = sides
  end

  def equilateral?
    a == b && b == c && legal?
  end

  def isosceles?
    (a == b || b == c || c == a) && legal?
  end

  def scalene?
    !isosceles? && !equilateral? && legal?
  end

  private

  def legal?
    !(a + b <= c || b + c <= a || c + a <= b)
  end
end

File: elixir/triangle/lib/triangle.ex
defmodule Triangle do
  def kind(a, b, c) when a <= 0 or b <= 0 or c <= 0 do
    {:error, "all side lengths must be positive"}
  end

  def kind(a, b, c) when a + b < c or b + c < a or a + c < b do
    {:error, "side lengths violate triangle inequality"}
  end

  def kind(a, b, c) when a == b and b == c do
    {:ok, :equilateral}
  end

  def kind(a, b, c) when a == b or b == c or a == c do
    {:ok, :isosceles}
  end

  def kind(_, _, _) do
    {:ok, :scalene}
  end
end
